src/doc/pythonbindings.rst

```diff
@@ -587,11 +587,16 @@ Section :ref:`sec-ImageSpec`, is replicated for Python.
 .. py:method:: ImageSpec.get_int_attribute (name, defaultval=0)
                   ImageSpec.get_float_attribute (name, defaultval=0.0)
                   ImageSpec.get_string_attribute (name, defaultval="")
+                  ImageSpec.get_bytes_attribute (name, defaultval="")
 
     Retrieves a named metadata value from `extra_attribs`, if it is
     found and is of the given type; returns the default value (or a passed
     value) if not found.
 
+    For an attribute of type STRING, get_bytes_attribute in Python3 skips
+    decoding the underlying C string as UTF-8 and returns a `bytes` object
+    containing the raw byte string.
+
```

Comment:
Thanks.

Just for my own understanding, how is the behavior different in Python 2? 

And exactly when would a user prefer to call get_bytes_attribute rather than get_string_attribute? Is it only after an exception was caught when calling get_string_attribute, indicating that the attribute is not value UTF-8?

Reply:
Is there merit to declaring that all string attributes in the spec or a ParamValueList MUST be UTF-8, and therefore the strings should be sanitized prior to being added to the list/spec in the first place?

Reply:
I'm python 2 there's already an if statement that defines `PY_STR` as `py::bytes` so `get_bytes_attribute` will behave the same as `get_string_attribute`

An example of how I'd use this in my python3 code, would be probably to always use `get_bytes_attribute` every time and deal with decoding myself (e.g. `str(b, encoding="utf-8", errors="ignore")`)

Perhaps this could be done instead by oiio under the hood, but I'm not sure if metadata values should always be forced to be utf-8. Maybe some values are intended to have some binary data.

With my example, I have an EXR with a metadata key `AudioInfo` and its value is `\xff\xff\xff...` repeated with `@` at the end. Clearly not particularly informative, and my ignoring decode errors leaves me with the string `"@"`. But should oiio be taking it upon itself to ditch all those `\xff`s and giving me just `"@"`?

Reply:
I think that it would be reasonable to require *string* metadata to be UTF-8. We already separately support metadata that's just an array of bytes, binary data (in fact, that's how we store ICCProfile metadata).

I'd need to look at the OpenEXR docs to see what they say about the contents of string metadata. I'm not sure they were thinking about UTF-8 at the time. But in general, a string isn't a good way to store arbitrary binary byte data, since a 0 character in the middle of a string will mess all sorts of things up.

Reply:
Hmmm, put another way: If you always retrieve strings as get_bytes_attribute and then do the str conversion... can you reliably distinguish between metadata that's already UTF-8, versus a binary blob that happens to be a sequence that is UTF-8 valid? Is there a danger of doing the conversion to UTF-8 one to many times, if you know what I mean?

Reply:
Ah, ok. If there's already a type for binary data, maybe this is unnecessary and oiio should sanitise string metadata values.

So is OpenEXR returning string values, but not caring about Unicode?

With my one example, I don't think the value is important. Could quite easily accept it's invalid and needs sanitising.

i.e. I'm not using that particular key. I'm just trying to avoid it erroring.

Reply:
Thinking about it a bit further, what's causing the error is `py::str` being used just in Python3 to convert the `char *`.

Therefore in Python2 the `char *`s just come out as is, there's no concept of converting from UTF-8.

I can see two options:

1. oiio in Python2 doesn't care about metadata string values being UTF-8 or not, so in Python3 it doesn't either, and all string metadata values are returned as `bytes`
2. oiio in Python3 now cares about metadata string values being UTF-8 and will sanitise them before returning as `str` - however, if you have some code relying on a metadata value that isn't UTF-8, you can still access the raw byte string using `get_bytes_attribute`

Reply:
I can see a byte array in oiio is done with `TypeDesc(TypeDesc::UINT8, total_length)`

However, looking at the OpenEXR code, I only see the attribute types:

`EXR_ATTR_STRING` and `EXR_ATTR_STRING_VECTOR`

None of them seem to define an arbitrary byte array.

Seeing as I have at least this one example of an EXR metadata value not being valid UTF-8, that suggests OpenEXR's definition of a string is just an array of bytes.

Therefore, in Python3 land (and I guess Python2 to be consistent) all EXR string metadata values would need to be `TypeDesc(TypeDesc::UINT8, total_length)` to truly represent their type, but that clearly wouldn't be backwards compatible and seems silly.

Are any EXR string metadata values ever intended to store binary data? They're not zero terminated strings, their length is specified and then the data. You could store any byte array in an EXR metadata string - I think.

As we have no way of knowing whether an EXR string metadata value is intended to be text or raw data, I guess best thing to do in Python3 is assume text, and diligently decode as UTF-8. In which case, this `get_bytes_attribute` provides a useful way to avoid decoding errors for code expecting a particular metadata key to contain raw data.

e.g. Naively I want to eye-ball some `CameraName` metadata key, and there it is in text full of emojis perhaps. However, I'm also specifically looking for a key called `IconPNG` which I know will contain the PNG data for a small icon. I'm in Python3 using a new version of oiio, so I know I can pull out this data using `get_bytes_attribute`

Reply:
I would like strings to be strings, if possible.

OIIO has a standard way of handling non-text data that's just a raw byte array, as you found. But OpenEXR does not (as far as I know; I've asked about this on the openexr mail list to verify). 

(Aside: I think the best approach is for OpenEXR to add a "byte array" attribute type, and declare string attributes to be UTF-8. But that will probably take some time and can't be 100% relied upon until a new OpenEXR release that supports this is widely adopted.)

I think it's an unnecessary burden on all python programs for OIIO to return bytes everyplace it now returns python str's, and expect the python script to explicitly convert to string. Conceptually, these are strings, so they should just return strings.

So here is what I propose:

For the python bindings, before returning a string from getattribute(), we should verify that it's valid UTF-8, and if it's not, then return it as a byte array rather than a string.

We can also keep your get_bytes_attribute, maybe there times when it's helpful to ask for that directly.

BTW, IconPNG -- do you really need it to be a fully encoded bit pattern of a PNG file? Or do you just need a preview image? Because there's already an OpenEXR "PreviewImage" attribute! (Though I think OIIO neglects to support it at the moment, but we could.)

Reply:
Oh, that `IconPNG` was just me reaching for an invented example of a reason to have binary data in an EXR metadata key